ast-checks/src/missing-patch-comment.rs

```diff
@@ -0,0 +1,100 @@
+use codespan::{FileId, Files};
+use comment_finders::{find_comment_above, find_comment_within};
+use rnix::{types::*};
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, env, error::Error, fs, process};
+use tree_utils::walk_keyvalues_filter_key;
+
+mod comment_finders;
+mod tree_utils;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct SourceLocation {
+    column: usize,
+    line: usize,
+    file: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct MissingCommentMessage {
+    cond: bool,
+    locations: Vec<SourceLocation>,
+    msg: &'static str,
+    name: &'static str,
+}
+
+fn main() -> Result<(), Box<dyn Error>> {
+    let args: Vec<String> = env::args().skip(1).collect();
+    let mut report = HashMap::new();
+
+    let mut files = Files::new();
+    for filename in args {
+        let file_id = files.add(filename.clone(), fs::read_to_string(&filename)?);
+        report.insert(filename.clone(), analyze_single_file(&files, file_id)?);
+    }
+
+    println!("{}", serde_json::to_string(&report)?);
+    Ok(())
+}
+
+fn analyze_single_file(
+    files: &Files<String>,
+    file_id: FileId,
+) -> Result<Vec<MissingCommentMessage>, Box<dyn Error>> {
+    let ast = rnix::parse(files.source(file_id)).as_result()?;
+    let root = ast.root().inner().ok_or("No elements")?;
+    let mut report: Vec<MissingCommentMessage> = vec![];
+
+    // Find a key in an attrset called 'patches'
+    for kv in walk_keyvalues_filter_key(&root, "patches") {
+        // Look for a comment directly above the `patches = ...` line
+        if let Some(_comment) = find_comment_above(kv.node()) {
+            // Okay!, we found a comment directly above this line,
+            // so I'm interpreting that as the comment which applies to
+            // all patches. It would be _better_ to put a single comment
+            // per patch, but we'll let this slide.
+            continue;
+        }
+
+        let value = kv.value().and_then(List::cast).unwrap_or_else(|| {
+            let offset = kv.node().text_range().start().to_usize() as u32;
+            let loc = files.location(file_id, offset).unwrap();
```

Comment:
@lovesegfault: I tried for ~30 minutes but I can't seem to get rid of this unwrap() and turn it into a ? without compiler errors. I tried lots of different ways, like changing the unwrap_or_else to ok_or_else, etc etc.

Reply:
It's nbd if this error formater panics, I guess.

Reply:
Interesting, let me play with it a bit.

Reply:
```
        let value = kv.value().and_then(List::cast).ok_or_else(|| {
            let offset = kv.node().text_range().start().to_usize() as u32;
            let loc = files.location(file_id, offset)?;
            eprintln!(
                "Unexpected format at line {} column {}. patches is not a list",
                loc.line, loc.column
            );
            process::exit(1);
        })?;
```
^ This is more of less what I _thought_ I should be able to convert it into, but this doesn't compile. I sort of understand why it doesn't...

This is attempt 2:
```
        let value = kv.value().and_then(List::cast).ok_or_else(|| {
            let offset = kv.node().text_range().start().to_usize() as u32;
            let loc = files.location(file_id, offset)?;
            Err(format!(
                "Unexpected format at line {} column {}. patches is not a list",
                loc.line, loc.column
            ))
        })?;
```
which doesn't compile for reasons I definitely don't understand (the trait `StdError` is not implemented for `std::result::Result<_, std::string::String>` and the trait `From<codespan_reporting::files::Error>` is not implemented for `std::string::String`)

Reply:
Okay, after thinking about this for a bit you have two choices.

1. Use a match like this:
```rust
        let value = match kv.value().and_then(List::cast) {
            Some(l) => l,
            None => {
                let offset = kv.node().text_range().start().to_usize() as u32;
                let loc = files.location(file_id, offset)?;
                eprintln!(
                    "Unexpected format at line {} column {}. patches is not a list",
                    loc.line, loc.column
                );
                process::exit(1);
            }
        };
```
2. Use an error handling library like `anyhow` and create an ad-hoc custom error type.

Given the amount of work put into this PR, the marginal gains, and the already good state of it I recommend you take the `match` route.

Hope this helps!

Reply:
Thanks! I took that suggestion.

Thanks for all your help.